compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,615 @@
+
+  // Step 5. Post-bufferization mapping to blocks and threads.
+  // Need to match again since bufferize invalidated all handles.
+  // TODO: assumes a single func::FuncOp to transform, may need hardening.
+  funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  funcH = mapToBlockAndThreads(b, funcH, infos.workgroupSize);
+
+  // Step 6. Post-bufferization vector distribution with rank-reduction.
+  distributeVectors(b, variantH, funcH);
+}
+
+static bool matchGPUReduction(linalg::LinalgOp op,
+                              GPUReductionStrategyInfos &info) {
+  // TODO: match the sequence the strategy supports.
+  StructuredOpMatcher pattern, fill, leadingEltwise, trailingEltwise;
+  makeGPUReductionMatcher(pattern, fill, leadingEltwise, trailingEltwise);
+  if (!matchPattern(op, pattern)) return false;
+
+  info.hasLeadingEltwise = leadingEltwise.getCaptured() != nullptr;
+  info.hasTrailingEltwise = trailingEltwise.getCaptured() != nullptr;
+
+  // Hardcoded workagroup size, this could be deduced from the reduction dim.
+  info.workgroupSize = {32, 2, 1};
+  SmallVector<unsigned> partitionedLoops =
+      cast<PartitionableLoopsInterface>(op.getOperation())
+          .getPartitionableLoops(kNumMaxParallelDims);
+  size_t numLoops = partitionedLoops.empty() ? 0 : partitionedLoops.back() + 1;
+  // Tile all the parallel dimension to 1.
+  info.workgroupTileSizes.append(numLoops, 1);
+  info.fillSecondTileSizes = {1, 0, 0};
+  info.genericSecondTileSizes = {1, 1, 0};
+  return true;
+}
+
+/// Structure to hold the parameters related to GPU reduction strategy.
+struct CPUReductionStrategyInfos {
+  int64_t workgroupSize;
+  SmallVector<int64_t> tileSizes;
+};
+
+static bool matchCPUReduction(linalg::LinalgOp op,
+                              CPUReductionStrategyInfos &infos) {
+  // TODO: match the sequence the strategy supports.
+  auto fill = m_StructuredOp<linalg::FillOp>();
+  auto pattern = m_StructuredOp()
+                     .dim(AllDims(), ShapeKind::Static)
+                     .dim(-1, utils::IteratorType::reduction)
+                     .output(NumEqualsTo(1))
+                     .output(0, fill);
+
+  // TODO: set the right config as expected by the strategy.
+  infos.workgroupSize = 1;
+  SmallVector<unsigned> partitionedLoops =
+      cast<PartitionableLoopsInterface>(op.getOperation())
+          .getPartitionableLoops(kNumMaxParallelDims);
+  size_t numLoops = partitionedLoops.empty() ? 0 : partitionedLoops.back() + 1;
+  // Tile all the parallel dimension to 1.
+  infos.tileSizes.append(numLoops, 1);
+  return true;
+}
+
+using StrategyBuilderFn = std::function<void(ImplicitLocOpBuilder &, Value)>;
+
+static void createTransformRegion(func::FuncOp entryPoint,
+                                  StrategyBuilderFn buildStrategy) {
+  MLIRContext *ctx = entryPoint.getContext();
+  Location loc = entryPoint.getLoc();
+  OpBuilder b(ctx);
+  auto mod = entryPoint->getParentOfType<ModuleOp>();
```

Comment:
This is inserting the transform dialect strategy within parent scope. So there is an implicit assumption that the pass itself is run on the parent scope. We can either pass the module in here explicitly to signify that the module is the translation unit, or we can insert it within the module, and discard it at the end of code-generation?

Reply:
Changed the code to follow your second suggestion, we now insert within the module and discard it in the dropSchedule pass.